File: EgsReport/CryptoService.vb
Imports System
Imports System.IO
Imports System.Text
Imports System.Security.Cryptography
Imports System.Numerics
Imports System.Globalization

''' <summary>
''' VB.NET translation of the provided crypto_service.ts
''' - AES-128-CBC with PKCS7 padding (Set/Set_Old, Get/Get_Old)
''' - HMAC-SHA256 -> hex -> BigInteger -> Base36 -> 25-char string (Get25CharString)
''' </summary>
Public Class CryptoService
    ' Provide the old environment secret here (used by Set_Old / Get_Old)
    Public Property EnvironmentSecret As String

    ' TS: set_old(value)
    Public Function Set_Old(value As String) As String
        If String.IsNullOrWhiteSpace(value) Then Return value
        If String.IsNullOrEmpty(EnvironmentSecret) Then
            Throw New InvalidOperationException("EnvironmentSecret is not set.")
        End If
        Dim key = Encoding.UTF8.GetBytes(EnvironmentSecret)
        Dim iv = Encoding.UTF8.GetBytes(EnvironmentSecret)
        ValidateKeyIv(key, iv)
        Return EncryptAesCbcPkcs7(value, key, iv)
    End Function

    ' TS: set(value) with fixed key/iv "4a2b4c99c0bc1012"
    Public Function [Set](value As String) As String
        If String.IsNullOrWhiteSpace(value) Then Return value
        Dim k As String = "4a2b4c99c0bc1012"
        Dim key = Encoding.UTF8.GetBytes(k)
        Dim iv = Encoding.UTF8.GetBytes(k)
        ValidateKeyIv(key, iv)
        Return EncryptAesCbcPkcs7(value, key, iv)
    End Function

    ' TS: get_old(value)
    Public Function Get_Old(encryptedBase64 As String) As String
        If String.IsNullOrEmpty(encryptedBase64) Then Return encryptedBase64
        If String.IsNullOrEmpty(EnvironmentSecret) Then
            Throw New InvalidOperationException("EnvironmentSecret is not set.")
        End If
        Dim key = Encoding.UTF8.GetBytes(EnvironmentSecret)
        Dim iv = Encoding.UTF8.GetBytes(EnvironmentSecret)
        ValidateKeyIv(key, iv)
        Return DecryptAesCbcPkcs7(encryptedBase64, key, iv)
    End Function

    ' TS: get(value) with fixed key/iv "4a2b4c99c0bc1012"
    Public Function [Get](encryptedBase64 As String) As String
        If String.IsNullOrEmpty(encryptedBase64) Then Return encryptedBase64
        Dim k As String = "4a2b4c99c0bc1012"
        Dim key = Encoding.UTF8.GetBytes(k)
        Dim iv = Encoding.UTF8.GetBytes(k)
        ValidateKeyIv(key, iv)
        Return DecryptAesCbcPkcs7(encryptedBase64, key, iv)
    End Function

    ' TS: get25CharString(code)
    ' HMAC-SHA256 with key "6v3b6d88c0bc1045", base36 uppercase, 25 chars (left-pad with '0' or take last 25)
    Public Shared Function Get25CharString(code As String) As String
        Dim hmacKey As Byte() = Encoding.UTF8.GetBytes("6v3b6d88c0bc1045")
        Dim data As Byte() = Encoding.UTF8.GetBytes(If(code, String.Empty))

        Dim hashBytes As Byte()
        Using hmac As New HMACSHA256(hmacKey)
            hashBytes = hmac.ComputeHash(data)
        End Using

        ' Convert to hex (lowercase is fine; NumberStyles.AllowHexSpecifier accepts both)
        Dim sb As New StringBuilder(hashBytes.Length * 2)
        For Each b In hashBytes
            sb.Append(b.ToString("x2"))
        Next
        Dim hex As String = sb.ToString()

        ' Convert hex string to BigInteger (treat as positive)
        Dim big As BigInteger = BigInteger.Parse(hex, NumberStyles.AllowHexSpecifier)
        If big.Sign < 0 Then big = BigInteger.Negate(big)

        ' Convert to base36 uppercase
        Dim base36 As String = ToBase36(big)
        If base36.Length = 0 Then base36 = "0"

        ' Ensure exactly 25 characters: left-pad with 0s or take last 25
        If base36.Length < 25 Then
            Return base36.PadLeft(25, "0"c)
        Else
            Return base36.Substring(base36.Length - 25)
        End If
    End Function

    ' -------------------- Internals --------------------

    Private Shared Sub ValidateKeyIv(key As Byte(), iv As Byte())
        ' AES-128 requires 16-byte key/iv
        If key Is Nothing OrElse key.Length <> 16 Then
            Throw New ArgumentException("Key must be 16 bytes (AES-128).")
        End If
        If iv Is Nothing OrElse iv.Length <> 16 Then
            Throw New ArgumentException("IV must be 16 bytes (AES-128).")
        End If
    End Sub

    Private Shared Function EncryptAesCbcPkcs7(plainText As String, key As Byte(), iv As Byte()) As String
        Using aes As Aes = Aes.Create()
            aes.Key = key
            aes.IV = iv
            aes.Mode = CipherMode.CBC
            aes.Padding = PaddingMode.PKCS7

            Using ms As New MemoryStream()
                Using cs As New CryptoStream(ms, aes.CreateEncryptor(), CryptoStreamMode.Write)
                    Dim inputBytes = Encoding.UTF8.GetBytes(plainText)
                    cs.Write(inputBytes, 0, inputBytes.Length)
                    cs.FlushFinalBlock()
                    Dim cipher As Byte() = ms.ToArray()
                    Return Convert.ToBase64String(cipher) ' CryptoJS AES.encrypt.toString() produces Base64 ciphertext
                End Using
            End Using
        End Using
    End Function

    Private Shared Function DecryptAesCbcPkcs7(cipherBase64 As String, key As Byte(), iv As Byte()) As String
        Dim cipherBytes = Convert.FromBase64String(cipherBase64)
        Using aes As Aes = Aes.Create()
            aes.Key = key
            aes.IV = iv
            aes.Mode = CipherMode.CBC
            aes.Padding = PaddingMode.PKCS7

            Using ms As New MemoryStream(cipherBytes)
                Using cs As New CryptoStream(ms, aes.CreateDecryptor(), CryptoStreamMode.Read)
                    Using sr As New StreamReader(cs, Encoding.UTF8)
                        Return sr.ReadToEnd()
                    End Using
                End Using
            End Using
        End Using
    End Function

    Private Shared Function ToBase36(value As BigInteger) As String
        If value.IsZero Then
            Return "0"
        End If

        Dim digits As Char() = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ".ToCharArray()
        Dim sb As New StringBuilder()
        Dim v As System.Numerics.BigInteger = System.Numerics.BigInteger.Abs(value)

        While v > System.Numerics.BigInteger.Zero
            Dim remainder As System.Numerics.BigInteger
            v = System.Numerics.BigInteger.DivRem(v, 36, remainder)
            ' remainder is guaranteed to fit into Integer (0..35)
            Dim idx As Integer = CInt(remainder)
            sb.Insert(0, digits(idx))
        End While

        Return sb.ToString()
    End Function
End Class

' Optional helper extension (not required by the class above)
Module StringExtensions
    <System.Runtime.CompilerServices.Extension>
    Public Function [Or](value As String, fallback As String) As String
        If String.IsNullOrEmpty(value) Then Return fallback
        Return value
    End Function
End Module

File: src/CmcStandardPrinting.Reporting/CryptoService.cs
using System;
using System.Globalization;
using System.IO;
using System.Numerics;
using System.Security.Cryptography;
using System.Text;

namespace CmcStandardPrinting.Reporting
{
    /// <summary>
    /// Legacy-compatible crypto helper that mirrors the VB implementation used by DevExpress reports.
    /// Supports AES-128-CBC encryption/decryption with PKCS7 padding and the legacy 25-character hash format.
    /// </summary>
    public sealed class CryptoService
    {
        /// <summary>
        /// Optional environment secret used by the legacy <c>Set_Old</c> and <c>Get_Old</c> flows.
        /// </summary>
        public string? EnvironmentSecret { get; set; }

        public string? SetOld(string? value)
        {
            if (string.IsNullOrWhiteSpace(value))
            {
                return value;
            }

            if (string.IsNullOrEmpty(EnvironmentSecret))
            {
                throw new InvalidOperationException("EnvironmentSecret is not set.");
            }

            var key = Encoding.UTF8.GetBytes(EnvironmentSecret);
            var iv = Encoding.UTF8.GetBytes(EnvironmentSecret);
            ValidateKeyAndIv(key, iv);
            return EncryptAesCbcPkcs7(value!, key, iv);
        }

        public string? Set(string? value)
        {
            if (string.IsNullOrWhiteSpace(value))
            {
                return value;
            }

            const string k = "4a2b4c99c0bc1012";
            var key = Encoding.UTF8.GetBytes(k);
            var iv = Encoding.UTF8.GetBytes(k);
            ValidateKeyAndIv(key, iv);
            return EncryptAesCbcPkcs7(value!, key, iv);
        }

        public string? GetOld(string? encryptedBase64)
        {
            if (string.IsNullOrEmpty(encryptedBase64))
            {
                return encryptedBase64;
            }

            if (string.IsNullOrEmpty(EnvironmentSecret))
            {
                throw new InvalidOperationException("EnvironmentSecret is not set.");
            }

            var key = Encoding.UTF8.GetBytes(EnvironmentSecret);
            var iv = Encoding.UTF8.GetBytes(EnvironmentSecret);
            ValidateKeyAndIv(key, iv);
            return DecryptAesCbcPkcs7(encryptedBase64!, key, iv);
        }

        public string? Get(string? encryptedBase64)
        {
            if (string.IsNullOrEmpty(encryptedBase64))
            {
                return encryptedBase64;
            }

            const string k = "4a2b4c99c0bc1012";
            var key = Encoding.UTF8.GetBytes(k);
            var iv = Encoding.UTF8.GetBytes(k);
            ValidateKeyAndIv(key, iv);
            return DecryptAesCbcPkcs7(encryptedBase64!, key, iv);
        }

        /// <summary>
        /// Generates the legacy 25-character hash string using HMAC-SHA256, hex, BigInteger, and base-36 steps.
        /// </summary>
        public static string Get25CharString(string? code)
        {
            var hmacKey = Encoding.UTF8.GetBytes("6v3b6d88c0bc1045");
            var data = Encoding.UTF8.GetBytes(code ?? string.Empty);

            byte[] hashBytes;
            using (var hmac = new HMACSHA256(hmacKey))
            {
                hashBytes = hmac.ComputeHash(data);
            }

            var sb = new StringBuilder(hashBytes.Length * 2);
            foreach (var b in hashBytes)
            {
                sb.Append(b.ToString("x2"));
            }

            var hex = sb.ToString();
            var big = BigInteger.Parse(hex, NumberStyles.AllowHexSpecifier);
            if (big.Sign < 0)
            {
                big = BigInteger.Negate(big);
            }

            var base36 = ToBase36(big);
            if (base36.Length == 0)
            {
                base36 = "0";
            }

            return base36.Length < 25
                ? base36.PadLeft(25, '0')
                : base36[^25..];
        }

        private static void ValidateKeyAndIv(byte[] key, byte[] iv)
        {
            if (key == null || key.Length != 16)
            {
                throw new ArgumentException("Key must be 16 bytes (AES-128).");
            }

            if (iv == null || iv.Length != 16)
            {
                throw new ArgumentException("IV must be 16 bytes (AES-128).");
            }
        }

        private static string EncryptAesCbcPkcs7(string plainText, byte[] key, byte[] iv)
        {
            using var aes = Aes.Create();
            aes.Key = key;
            aes.IV = iv;
            aes.Mode = CipherMode.CBC;
            aes.Padding = PaddingMode.PKCS7;

            using var ms = new MemoryStream();
            using (var cs = new CryptoStream(ms, aes.CreateEncryptor(), CryptoStreamMode.Write))
            {
                var inputBytes = Encoding.UTF8.GetBytes(plainText);
                cs.Write(inputBytes, 0, inputBytes.Length);
                cs.FlushFinalBlock();
            }

            var cipher = ms.ToArray();
            return Convert.ToBase64String(cipher);
        }

        private static string DecryptAesCbcPkcs7(string cipherBase64, byte[] key, byte[] iv)
        {
            var cipherBytes = Convert.FromBase64String(cipherBase64);
            using var aes = Aes.Create();
            aes.Key = key;
            aes.IV = iv;
            aes.Mode = CipherMode.CBC;
            aes.Padding = PaddingMode.PKCS7;

            using var ms = new MemoryStream(cipherBytes);
            using var cs = new CryptoStream(ms, aes.CreateDecryptor(), CryptoStreamMode.Read);
            using var sr = new StreamReader(cs, Encoding.UTF8);
            return sr.ReadToEnd();
        }

        private static string ToBase36(BigInteger value)
        {
            if (value.IsZero)
            {
                return "0";
            }

            const string digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
            var sb = new StringBuilder();
            var v = BigInteger.Abs(value);

            while (v > BigInteger.Zero)
            {
                v = BigInteger.DivRem(v, 36, out var remainder);
                var idx = (int)remainder;
                sb.Insert(0, digits[idx]);
            }

            return sb.ToString();
        }
    }
}
